src/signalrclient/signalr_default_scheduler.cpp

```diff
@@ -0,0 +1,217 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "stdafx.h"
+#include <assert.h>
+#include "signalr_default_scheduler.h"
+#include <thread>
+
+namespace signalr
+{
+    thread::thread()
+        : m_internals(std::make_shared<internals>())
+    {
+        auto internals = m_internals;
+
+        m_thread = std::thread([=]()
+            {
+                while (true)
+                {
+                    {
+                        signalr_base_cb cb;
+                        {
+                            std::unique_lock<std::mutex> lock(internals->m_callback_lock);
+                            auto& callback = internals->m_callback;
+                            auto& closed = internals->m_closed;
+
+                            if (closed && callback == nullptr)
+                            {
+                                assert(callback == nullptr);
+                                return;
+                            }
+
+                            if (callback == nullptr)
+                            {
+                                internals->m_callback_cv.wait(lock,
+                                    [&callback, &closed]
+                                    {
+                                        return closed || callback != nullptr;
+                                    });
+                            }
+
+                            if (closed && callback == nullptr)
+                            {
+                                assert(callback == nullptr);
+                                return;
+                            }
+
+                            cb = callback;
+                            internals->m_callback = nullptr;
```

Comment:
Is m_callback only ever set one at a time? I don't know a ton about these callback mechanisms, but can there be a chain of callbacks?

Reply:
This one is only ever one at a time.